File: hfo2/src/manifest.rs
/*
 * Copyright 2019 Sanguk Park
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use core::convert::TryInto;
use core::fmt::{self, Write};
use core::mem::MaybeUninit;
use core::ptr;

use crate::fdt::*;
use crate::memiter::*;
use crate::types::*;

use arrayvec::ArrayVec;

const VM_NAME_BUF_SIZE: usize = 2 + 5 + 1; // "vm" + number + null terminator
const_assert!(MAX_VMS <= 99999);

#[derive(PartialEq, Debug)]
pub enum Error {
    NoHypervisorFdtNode,
    NotCompatible,
    ReservedVmId,
    NoPrimaryVm,
    TooManyVms,
    PropertyNotFound,
    MalformedString,
    StringTooLong,
    MalformedStringList,
    MalformedInteger,
    IntegerOverflow,
}

impl Into<&'static str> for Error {
    fn into(self) -> &'static str {
        use Error::*;
        match self {
            NoHypervisorFdtNode => "Could not find \"hypervisor\" node in manifest",
            NotCompatible => "Hypervisor manifest entry not compatible with Hafnium",
            ReservedVmId => "Manifest defines a VM with a reserved ID",
            NoPrimaryVm => "Manifest does not contain a primary VM entry",
            TooManyVms => {
                "Manifest specifies more VMs than Hafnium has statically allocated space for"
            }
            PropertyNotFound => "Property not found",
            MalformedString => "Malformed string property",
            StringTooLong => "String too long",
            MalformedStringList => "Malformed string list property",
            MalformedInteger => "Malformed integer property",
            IntegerOverflow => "Integer overflow",
        }
    }
}

/// Maximum length of a string parsed from the FDT, including NULL terminator.
const MANIFEST_MAX_STRING_LENGTH: usize = 32;

/// Holds information about one of the VMs described in the manifest.
#[derive(Debug)]
pub struct ManifestVm {
    // Properties defined for both primary and secondary VMs.
    pub debug_name: [u8; MANIFEST_MAX_STRING_LENGTH],

    // Properties specific to secondary VMs.
    pub kernel_filename: [u8; MANIFEST_MAX_STRING_LENGTH],
    pub mem_size: u64,
    pub vcpu_count: spci_vcpu_count_t,
}

/// Hafnium manifest parsed from FDT.
#[derive(Debug)]
pub struct Manifest {
    pub vms: ArrayVec<[ManifestVm; MAX_VMS]>,
}

/// Generates a string with the two letters "vm" followed by an integer.
fn generate_vm_node_name<'a>(
    buf: &'a mut [u8; VM_NAME_BUF_SIZE],
    vm_id: spci_vm_id_t,
) -> &'a mut [u8] {
    struct BufWrite<'a> {
        buf: &'a mut [u8; VM_NAME_BUF_SIZE],
        size: usize,
    }

    impl<'a> Write for BufWrite<'a> {
        fn write_str(&mut self, s: &str) -> Result<(), fmt::Error> {
            let dest = self
                .buf
                .get_mut(self.size..(self.size + s.len()))
                .ok_or(fmt::Error)?;
            dest.copy_from_slice(s.as_bytes());
            self.size += s.len();

            Ok(())
        }
    }

    let mut buf = BufWrite { buf, size: 0 };
    write!(buf, "vm{}\0", vm_id).unwrap();
    &mut buf.buf[..buf.size]
}

impl<'a> FdtNode<'a> {
    /// TODO(HfO2): This function is marked `inline(never)`, to prevent stack overflow. It is still
    /// mysterious why inlining this function into ManifestVm::new makes stack overflow.
    #[inline(never)]
    fn read_string(&self, property: *const u8, out: &mut [u8]) -> Result<(), Error> {
        let data = self
            .read_property(property)
            .map_err(|_| Error::PropertyNotFound)?;

        // Require that the value contains exactly one NULL character and that it is the last byte.
        if data.iter().position(|&c| c == b'\0') != Some(data.len() - 1) {
            return Err(Error::MalformedString);
        }

        // Check that the string fits into the buffer.
        if data.len() > out.len() {
            return Err(Error::StringTooLong);
        }

        out[..data.len()].copy_from_slice(data);
        Ok(())
    }

    #[inline(never)]
    fn read_u64(&self, property: *const u8) -> Result<u64, Error> {
        let data = self
            .read_property(property)
            .map_err(|_| Error::PropertyNotFound)?;

        fdt_parse_number(data).ok_or(Error::MalformedInteger)
    }

    #[inline(never)]
    fn read_u16(&self, property: *const u8) -> Result<u16, Error> {
        let value = self.read_u64(property)?;

        value.try_into().map_err(|_| Error::IntegerOverflow)
    }
}

/// Represents the value of property whose type is a list of strings. These are encoded as one
/// contiguous byte buffer with NULL-separated entries.
#[derive(Clone)]
struct StringList {
    mem_it: MemIter,
}

impl StringList {
    fn read_from<'a>(node: &FdtNode<'a>, property: *const u8) -> Result<Self, Error> {
        let data = node
            .read_property(property)
            .map_err(|_| Error::PropertyNotFound)?;

        // Require that the value ends with a NULL terminator. Other NULL characters separate the
        // string list entries.
        if *data.last().unwrap() != b'\0' {
            return Err(Error::MalformedStringList);
        }

        Ok(Self {
            mem_it: unsafe { MemIter::from_raw(data.as_ptr(), data.len() - 1) },
        })
    }

    fn has_next(&self) -> bool {
        self.mem_it.len() > 0
    }

    fn get_next(&mut self) -> MemIter {
        assert!(self.has_next());

        let null_term = unsafe { self.mem_it.as_slice() }
            .iter()
            .position(|&c| c == b'\0');
        if let Some(pos) = null_term {
            // Found NULL terminator. Set entry memiter to byte range [base, null) and move list
            // memiter past the terminator.
            let ret = unsafe { MemIter::from_raw(self.mem_it.get_next(), pos) };
            self.mem_it.advance(pos + 1).unwrap();
            ret
        } else {
            // NULL terminator not found, this is the last entry.
            // Set entry memiter to the entire byte range and advance list memiter to the end of
            // the byte range.
            let ret = self.mem_it.clone();
            self.mem_it.advance(self.mem_it.len()).unwrap();
            ret
        }
    }

    fn contains(&self, asciz: &[u8]) -> bool {
        let mut it = self.clone();

        while it.has_next() {
            let entry = it.get_next();
            if unsafe { entry.iseq(asciz.as_ptr()) } {
                return true;
            }
        }

        false
    }
}

impl ManifestVm {
    fn new<'a>(node: &FdtNode<'a>, vm_id: spci_vm_id_t) -> Result<Self, Error> {
        let mut debug_name: [u8; MANIFEST_MAX_STRING_LENGTH] = Default::default();
        node.read_string("debug_name\0".as_ptr(), &mut debug_name)?;

        let mut kernel_filename: [u8; MANIFEST_MAX_STRING_LENGTH] = Default::default();

        let (mem_size, vcpu_count) = if vm_id != HF_PRIMARY_VM_ID {
            node.read_string("kernel_filename\0".as_ptr(), &mut kernel_filename)?;
            (
                node.read_u64("mem_size\0".as_ptr())?,
                node.read_u16("vcpu_count\0".as_ptr())?,
            )
        } else {
            (0, 0)
        };

        Ok(Self {
            debug_name,
            kernel_filename,
            mem_size,
            vcpu_count,
        })
    }
}

impl Manifest {
    /// Parse manifest from FDT.
    pub fn init<'a>(&mut self, fdt: &FdtNode<'a>) -> Result<(), Error> {
        let mut vm_name_buf = Default::default();
        let mut found_primary_vm = false;
        unsafe {
            self.vms.set_len(0);
        }

        // Find hypervisor node.
        let mut hyp_node = fdt.clone();
        hyp_node
            .find_child("hypervisor\0".as_ptr())
            .ok_or(Error::NoHypervisorFdtNode)?;

        // Check "compatible" property.
        let compatible_list = StringList::read_from(&hyp_node, "compatible\0".as_ptr())?;
        if !compatible_list.contains(b"hafnium,hafnium\0") {
            return Err(Error::NotCompatible);
        }

        // Iterate over reserved VM IDs and check no such nodes exist.
        for vm_id in 0..HF_VM_ID_OFFSET {
            let mut vm_node = hyp_node.clone();
            let vm_name = generate_vm_node_name(&mut vm_name_buf, vm_id);

            if vm_node.find_child(vm_name.as_ptr()).is_some() {
                return Err(Error::ReservedVmId);
            }
        }

        // Iterate over VM nodes until we find one that does not exist.
        for i in 0..=MAX_VMS as spci_vm_id_t {
            let vm_id = HF_VM_ID_OFFSET + i;
            let mut vm_node = hyp_node.clone();
            let vm_name = generate_vm_node_name(&mut vm_name_buf, vm_id);

            if vm_node.find_child(vm_name.as_ptr()).is_none() {
                break;
            }

            if i == MAX_VMS as spci_vm_id_t {
                return Err(Error::TooManyVms);
            }

            if vm_id == HF_PRIMARY_VM_ID {
                assert!(found_primary_vm == false); // sanity check
                found_primary_vm = true;
            }

            self.vms.push(ManifestVm::new(&vm_node, vm_id)?);
        }

        if !found_primary_vm {
            Err(Error::NoPrimaryVm)
        } else {
            Ok(())
        }
    }
}

#[cfg(test)]
mod test {
    extern crate std;
    use std::fmt::Write as fmtWrite;
    use std::io::Write;
    use std::mem;
    use std::mem::MaybeUninit;
    use std::process::*;
    use std::string::String;
    use std::vec::Vec;

    use super::*;
    use crate::utils::*;

    /// Class for programatically building a Device Tree.
    ///
    /// # Usage
    /// ```
    /// let dtb = ManifestDtBuilder::new()
    ///     .Command1()
    ///     .Command2()
    ///     ...
    ///     .CommandN()
    ///     .Build();
    /// ```
    struct ManifestDtBuilder {
        dts: String,
    }

    impl ManifestDtBuilder {
        fn new() -> Self {
            let mut builder = Self { dts: String::new() };
            builder.dts.push_str("/dts-v1/;\n");
            builder.dts.push_str("\n");

            // Start root node.
            builder.start_child("/");
            builder
        }

        fn build(&mut self) -> Vec<u8> {
            self.end_child();
            const program: &'static str = "../build/image/dtc.py";

            let mut child = Command::new(program)
                .stdin(Stdio::piped())
                .stdout(Stdio::piped())
                .args(&["compile"])
                .spawn()
                .unwrap();

            child
                .stdin
                .as_mut()
                .unwrap()
                .write_all(self.dts.as_bytes())
                .unwrap();

            child.wait_with_output().unwrap().stdout
        }

        fn start_child(&mut self, name: &str) -> &mut Self {
            self.dts.push_str(name);
            self.dts.push_str(" {\n");
            self
        }

        fn end_child(&mut self) -> &mut Self {
            self.dts.push_str("};\n");
            self
        }

        fn compatible(&mut self, value: &[&str]) -> &mut Self {
            self.string_list_property("compatible", value)
        }

        fn compatible_hafnium(&mut self) -> &mut Self {
            self.compatible(&["hafnium,hafnium"])
        }

        fn debug_name(&mut self, value: &str) -> &mut Self {
            self.string_property("debug_name", value)
        }

        fn kernel_filename(&mut self, value: &str) -> &mut Self {
            self.string_property("kernel_filename", value)
        }

        fn vcpu_count(&mut self, value: u64) -> &mut Self {
            self.integer_property("vcpu_count", value)
        }

        fn mem_size(&mut self, value: u64) -> &mut Self {
            self.integer_property("mem_size", value)
        }

        fn string_property(&mut self, name: &str, value: &str) -> &mut Self {
            write!(self.dts, "{} = \"{}\";\n", name, value).unwrap();
            self
        }

        fn string_list_property(&mut self, name: &str, value: &[&str]) -> &mut Self {
            write!(self.dts, "{} = \"", name);
            self.dts.push_str(&value.join("\", \""));
            self.dts.push_str("\";\n");
            self
        }

        fn integer_property(&mut self, name: &str, value: u64) -> &mut Self {
            write!(self.dts, "{} = <{}>;\n", name, value).unwrap();
            self
        }
    }

    fn get_fdt_root<'a>(dtb: &'a [u8]) -> Option<FdtNode<'a>> {
        let fdt_header = unsafe { &*(dtb.as_ptr() as *const FdtHeader) };

        let mut node = FdtNode::new_root(fdt_header)?;
        node.find_child("\0".as_ptr());
        Some(node)
    }

    #[test]
    fn no_hypervisor_node() {
        let dtb = ManifestDtBuilder::new().build();

        let fdt_root = get_fdt_root(&dtb).unwrap();
        let mut m: Manifest = unsafe { MaybeUninit::uninit().assume_init() };
        assert_eq!(m.init(&fdt_root).unwrap_err(), Error::NoHypervisorFdtNode);
    }

    #[test]
    fn no_compatible_property() {
        let dtb = ManifestDtBuilder::new()
            .start_child("hypervisor")
            .end_child()
            .build();

        let fdt_root = get_fdt_root(&dtb).unwrap();
        let mut m: Manifest = unsafe { MaybeUninit::uninit().assume_init() };
        assert_eq!(m.init(&fdt_root).unwrap_err(), Error::PropertyNotFound);
    }

    #[test]
    fn not_compatible() {
        let dtb = ManifestDtBuilder::new()
            .start_child("hypervisor")
            .compatible(&["foo,bar"])
            .end_child()
            .build();

        let fdt_root = get_fdt_root(&dtb).unwrap();
        let mut m: Manifest = unsafe { MaybeUninit::uninit().assume_init() };
        assert_eq!(m.init(&fdt_root).unwrap_err(), Error::NotCompatible);
    }

    #[test]
    fn compatible_one_of_many() {
        let dtb = ManifestDtBuilder::new()
            .start_child("hypervisor")
            .compatible(&["foo,bar", "hafnium,hafnium"])
            .start_child("vm1")
            .debug_name("primary")
            .end_child()
            .end_child()
            .build();

        let fdt_root = get_fdt_root(&dtb).unwrap();
        let mut m: Manifest = unsafe { MaybeUninit::uninit().assume_init() };
        m.init(&fdt_root).unwrap();
    }

    #[test]
    fn no_vm_nodes() {
        let dtb = ManifestDtBuilder::new()
            .start_child("hypervisor")
            .compatible_hafnium()
            .end_child()
            .build();

        let fdt_root = get_fdt_root(&dtb).unwrap();
        let mut m: Manifest = unsafe { MaybeUninit::uninit().assume_init() };
        assert_eq!(m.init(&fdt_root).unwrap_err(), Error::NoPrimaryVm);
    }

    #[test]
    fn long_string() {
        fn gen_long_string_dtb(valid: bool) -> Vec<u8> {
            const LAST_VALID: &'static str = "1234567890123456789012345678901";
            const FIRST_INVALID: &'static str = "12345678901234567890123456789012";
            assert_eq!(LAST_VALID.len() + 1, MANIFEST_MAX_STRING_LENGTH);
            assert_eq!(FIRST_INVALID.len() + 1, MANIFEST_MAX_STRING_LENGTH + 1);

            ManifestDtBuilder::new()
                .start_child("hypervisor")
                .compatible_hafnium()
                .start_child("vm1")
                .debug_name(if valid { LAST_VALID } else { FIRST_INVALID })
                .end_child()
                .end_child()
                .build()
        }

        let dtb_last_valid = gen_long_string_dtb(true);
        let dtb_first_invalid = gen_long_string_dtb(false);

        let fdt_root = get_fdt_root(&dtb_last_valid).unwrap();
        let mut m: Manifest = unsafe { MaybeUninit::uninit().assume_init() };
        m.init(&fdt_root).unwrap();

        let fdt_root = get_fdt_root(&dtb_first_invalid).unwrap();
        assert_eq!(m.init(&fdt_root).unwrap_err(), Error::StringTooLong);
    }

    #[test]
    fn reserved_vm_id() {
        let dtb = ManifestDtBuilder::new()
            .start_child("hypervisor")
            .compatible_hafnium()
            .start_child("vm1")
            .debug_name("primary_vm")
            .end_child()
            .start_child("vm0")
            .debug_name("reserved_vm")
            .vcpu_count(1)
            .mem_size(0x1000)
            .kernel_filename("kernel")
            .end_child()
            .end_child()
            .build();

        let fdt_root = get_fdt_root(&dtb).unwrap();
        let mut m: Manifest = unsafe { MaybeUninit::uninit().assume_init() };
        assert_eq!(m.init(&fdt_root).unwrap_err(), Error::ReservedVmId);
    }

    #[test]
    fn vcpu_count_limit() {
        fn gen_vcpu_count_limit_dtb(vcpu_count: u64) -> Vec<u8> {
            ManifestDtBuilder::new()
                .start_child("hypervisor")
                .compatible_hafnium()
                .start_child("vm1")
                .debug_name("primary_vm")
                .end_child()
                .start_child("vm2")
                .debug_name("secondary_vm")
                .vcpu_count(vcpu_count)
                .mem_size(0x1000)
                .kernel_filename("kernel")
                .end_child()
                .end_child()
                .build()
        }

        let dtb_last_valid = gen_vcpu_count_limit_dtb(u16::max_value() as u64);
        let dtb_first_invalid = gen_vcpu_count_limit_dtb(u16::max_value() as u64 + 1);

        let fdt_root = get_fdt_root(&dtb_last_valid).unwrap();
        let mut m: Manifest = unsafe { MaybeUninit::uninit().assume_init() };
        m.init(&fdt_root).unwrap();
        assert_eq!(m.vms.len(), 2);
        assert_eq!(m.vms[1].vcpu_count, u16::max_value());

        let fdt_root = get_fdt_root(&dtb_first_invalid).unwrap();
        assert_eq!(m.init(&fdt_root).unwrap_err(), Error::IntegerOverflow);
    }

    #[test]
    fn valid() {
        let dtb = ManifestDtBuilder::new()
            .start_child("hypervisor")
            .compatible_hafnium()
            .start_child("vm1")
            .debug_name("primary_vm")
            .end_child()
            .start_child("vm3")
            .debug_name("second_secondary_vm")
            .vcpu_count(43)
            .mem_size(0x12345)
            .kernel_filename("second_kernel")
            .end_child()
            .start_child("vm2")
            .debug_name("first_secondary_vm")
            .vcpu_count(42)
            .mem_size(12345)
            .kernel_filename("first_kernel")
            .end_child()
            .end_child()
            .build();

        let fdt_root = get_fdt_root(&dtb).unwrap();
        let mut m: Manifest = unsafe { MaybeUninit::uninit().assume_init() };
        m.init(&fdt_root).unwrap();
        assert_eq!(m.vms.len(), 3);

        let vm = &m.vms[0];
        assert_eq!(as_asciz(&vm.debug_name), b"primary_vm");

        let vm = &m.vms[1];
        assert_eq!(as_asciz(&vm.debug_name), b"first_secondary_vm");
        assert_eq!(vm.vcpu_count, 42);
        assert_eq!(vm.mem_size, 12345);
        assert_eq!(as_asciz(&vm.kernel_filename), b"first_kernel");

        let vm = &m.vms[2];
        assert_eq!(as_asciz(&vm.debug_name), b"second_secondary_vm");
        assert_eq!(vm.vcpu_count, 43);
        assert_eq!(vm.mem_size, 0x12345);
        assert_eq!(as_asciz(&vm.kernel_filename), b"second_kernel");
    }
}

File: experiments/ownership-inference/hafnium/src/manifest.c
/*
 * Copyright 2019 The Hafnium Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "hf/manifest.h"

#include "hf/addr.h"
#include "hf/check.h"
#include "hf/dlog.h"
#include "hf/fdt.h"
#include "hf/static_assert.h"
#include "hf/std.h"

#define TRY(expr)                                            \
	do {                                                 \
		enum manifest_return_code ret_code = (expr); \
		if (ret_code != MANIFEST_SUCCESS) {          \
			return ret_code;                     \
		}                                            \
	} while (0)

#define VM_NAME_BUF_SIZE (2 + 5 + 1) /* "vm" + number + null terminator */
static_assert(MAX_VMS <= 99999, "Insufficient VM_NAME_BUF_SIZE");
static_assert(HF_TEE_VM_ID > MAX_VMS,
	      "TrustZone VM ID clashes with normal VM range.");

/**
 * Generates a string with the two letters "vm" followed by an integer.
 * Assumes `buf` is of size VM_NAME_BUF_SIZE.
 */
static const char *generate_vm_node_name(char *buf, spci_vm_id_t vm_id)
{
	static const char *digits = "0123456789";
	char *ptr = buf + VM_NAME_BUF_SIZE;

	*(--ptr) = '\0';
	do {
		*(--ptr) = digits[vm_id % 10];
		vm_id /= 10;
	} while (vm_id);
	*(--ptr) = 'm';
	*(--ptr) = 'v';

	return ptr;
}

/**
 * Read a boolean property: true if present; false if not. If present, the value
 * of the property must be empty else it is considered malformed.
 */
static enum manifest_return_code read_bool(const struct fdt_node *node,
					   const char *property, bool *out)
{
	const char *data;
	uint32_t size;
	bool present = fdt_read_property(node, property, &data, &size);

	if (present && size != 0) {
		return MANIFEST_ERROR_MALFORMED_BOOLEAN;
	}

	*out = present;
	return MANIFEST_SUCCESS;
}

static enum manifest_return_code read_string(const struct fdt_node *node,
					     const char *property,
					     struct string *out)
{
	const char *data;
	uint32_t size;

	if (!fdt_read_property(node, property, &data, &size)) {
		return MANIFEST_ERROR_PROPERTY_NOT_FOUND;
	}

	switch (string_init(out, data, size)) {
	case STRING_SUCCESS:
		return MANIFEST_SUCCESS;
	case STRING_ERROR_INVALID_INPUT:
		return MANIFEST_ERROR_MALFORMED_STRING;
	case STRING_ERROR_TOO_LONG:
		return MANIFEST_ERROR_STRING_TOO_LONG;
	}
}

static enum manifest_return_code read_optional_string(
	const struct fdt_node *node, const char *property, struct string *out)
{
	enum manifest_return_code ret;

	ret = read_string(node, property, out);
	if (ret == MANIFEST_ERROR_PROPERTY_NOT_FOUND) {
		string_init_empty(out);
		ret = MANIFEST_SUCCESS;
	}
	return ret;
}

static enum manifest_return_code read_uint64(const struct fdt_node *node,
					     const char *property,
					     uint64_t *out)
{
	const char *data;
	uint32_t size;

	if (!fdt_read_property(node, property, &data, &size)) {
		return MANIFEST_ERROR_PROPERTY_NOT_FOUND;
	}

	if (!fdt_parse_number(data, size, out)) {
		return MANIFEST_ERROR_MALFORMED_INTEGER;
	}

	return MANIFEST_SUCCESS;
}

static enum manifest_return_code read_uint16(const struct fdt_node *node,
					     const char *property,
					     uint16_t *out)
{
	uint64_t value;

	TRY(read_uint64(node, property, &value));

	if (value > UINT16_MAX) {
		return MANIFEST_ERROR_INTEGER_OVERFLOW;
	}

	*out = (uint16_t)value;
	return MANIFEST_SUCCESS;
}

struct uint32list_iter {
	struct memiter mem_it;
};

static enum manifest_return_code read_optional_uint32list(
	const struct fdt_node *node, const char *property,
	struct uint32list_iter *out)
{
	const char *data;
	uint32_t size;

	if (!fdt_read_property(node, property, &data, &size)) {
		memiter_init(&out->mem_it, NULL, 0);
		return MANIFEST_SUCCESS;
	}

	if ((size % sizeof(uint32_t)) != 0) {
		return MANIFEST_ERROR_MALFORMED_INTEGER_LIST;
	}

	memiter_init(&out->mem_it, data, size);
	return MANIFEST_SUCCESS;
}

/**
 * Represents the value of property whose type is a list of strings. These are
 * encoded as one contiguous byte buffer with NULL-separated entries.
 */
struct stringlist_iter {
	struct memiter mem_it;
};

static enum manifest_return_code read_stringlist(const struct fdt_node *node,
						 const char *property,
						 struct stringlist_iter *out)
{
	const char *data;
	uint32_t size;

	if (!fdt_read_property(node, property, &data, &size)) {
		return MANIFEST_ERROR_PROPERTY_NOT_FOUND;
	}

	/*
	 * Require that the value ends with a NULL terminator. Other NULL
	 * characters separate the string list entries.
	 */
	if (data[size - 1] != '\0') {
		return MANIFEST_ERROR_MALFORMED_STRING_LIST;
	}

	memiter_init(&out->mem_it, data, size - 1);
	return MANIFEST_SUCCESS;
}

static bool uint32list_has_next(const struct uint32list_iter *list)
{
	return memiter_size(&list->mem_it) > 0;
}

static uint32_t uint32list_get_next(struct uint32list_iter *list)
{
	const char *mem_base = memiter_base(&list->mem_it);
	uint64_t num;

	CHECK(uint32list_has_next(list));

	if (!fdt_parse_number(mem_base, sizeof(uint32_t), &num)) {
		return MANIFEST_ERROR_MALFORMED_INTEGER;
	}

	memiter_advance(&list->mem_it, sizeof(uint32_t));
	return num;
}

static bool stringlist_has_next(const struct stringlist_iter *list)
{
	return memiter_size(&list->mem_it) > 0;
}

static void stringlist_get_next(struct stringlist_iter *list,
				struct memiter *out)
{
	const char *mem_base = memiter_base(&list->mem_it);
	size_t mem_size = memiter_size(&list->mem_it);
	const char *null_term;

	CHECK(stringlist_has_next(list));

	null_term = memchr(mem_base, '\0', mem_size);
	if (null_term == NULL) {
		/*
		 * NULL terminator not found, this is the last entry.
		 * Set entry memiter to the entire byte range and advance list
		 * memiter to the end of the byte range.
		 */
		memiter_init(out, mem_base, mem_size);
		memiter_advance(&list->mem_it, mem_size);
	} else {
		/*
		 * Found NULL terminator. Set entry memiter to byte range
		 * [base, null) and move list memiter past the terminator.
		 */
		size_t entry_size = null_term - mem_base;

		memiter_init(out, mem_base, entry_size);
		memiter_advance(&list->mem_it, entry_size + 1);
	}
}

static bool stringlist_contains(const struct stringlist_iter *list,
				const char *str)
{
	struct stringlist_iter it = *list;
	struct memiter entry;

	while (stringlist_has_next(&it)) {
		stringlist_get_next(&it, &entry);
		if (memiter_iseq(&entry, str)) {
			return true;
		}
	}
	return false;
}

static enum manifest_return_code parse_vm(struct fdt_node *node,
					  struct manifest_vm *vm,
					  spci_vm_id_t vm_id)
{
	struct uint32list_iter smcs;

	TRY(read_string(node, "debug_name", &vm->debug_name));
	TRY(read_optional_string(node, "kernel_filename",
				 &vm->kernel_filename));

	TRY(read_optional_uint32list(node, "smc_whitelist", &smcs));
	while (uint32list_has_next(&smcs) &&
	       vm->smc_whitelist.smc_count < MAX_SMCS) {
		vm->smc_whitelist.smcs[vm->smc_whitelist.smc_count++] =
			uint32list_get_next(&smcs);
	}

	if (uint32list_has_next(&smcs)) {
		dlog("%s SMC whitelist too long.\n", vm->debug_name);
	}

	TRY(read_bool(node, "smc_whitelist_permissive",
		      &vm->smc_whitelist.permissive));

	if (vm_id == HF_PRIMARY_VM_ID) {
		TRY(read_optional_string(node, "ramdisk_filename",
					 &vm->primary.ramdisk_filename));
	} else {
		TRY(read_uint64(node, "mem_size", &vm->secondary.mem_size));
		TRY(read_uint16(node, "vcpu_count", &vm->secondary.vcpu_count));
	}
	return MANIFEST_SUCCESS;
}

/**
 * Parse manifest from FDT.
 */
enum manifest_return_code manifest_init(struct manifest *manifest,
					const struct fdt_node *fdt_root)
{
	char vm_name_buf[VM_NAME_BUF_SIZE];
	struct fdt_node hyp_node;
	struct stringlist_iter compatible_list;
	size_t i = 0;
	bool found_primary_vm = false;

	memset_s(manifest, sizeof(*manifest), 0, sizeof(*manifest));

	/* Find hypervisor node. */
	hyp_node = *fdt_root;
	if (!fdt_find_child(&hyp_node, "hypervisor")) {
		return MANIFEST_ERROR_NO_HYPERVISOR_FDT_NODE;
	}

	/* Check "compatible" property. */
	TRY(read_stringlist(&hyp_node, "compatible", &compatible_list));
	if (!stringlist_contains(&compatible_list, "hafnium,hafnium")) {
		return MANIFEST_ERROR_NOT_COMPATIBLE;
	}

	/* Iterate over reserved VM IDs and check no such nodes exist. */
	for (i = 0; i < HF_VM_ID_OFFSET; i++) {
		spci_vm_id_t vm_id = (spci_vm_id_t)i;
		struct fdt_node vm_node = hyp_node;
		const char *vm_name = generate_vm_node_name(vm_name_buf, vm_id);

		if (fdt_find_child(&vm_node, vm_name)) {
			return MANIFEST_ERROR_RESERVED_VM_ID;
		}
	}

	/* Iterate over VM nodes until we find one that does not exist. */
	for (i = 0; i <= MAX_VMS; ++i) {
		spci_vm_id_t vm_id = HF_VM_ID_OFFSET + i;
		struct fdt_node vm_node = hyp_node;
		const char *vm_name = generate_vm_node_name(vm_name_buf, vm_id);

		if (!fdt_find_child(&vm_node, vm_name)) {
			break;
		}

		if (i == MAX_VMS) {
			return MANIFEST_ERROR_TOO_MANY_VMS;
		}

		if (vm_id == HF_PRIMARY_VM_ID) {
			CHECK(found_primary_vm == false); /* sanity check */
			found_primary_vm = true;
		}

		manifest->vm_count = i + 1;
		TRY(parse_vm(&vm_node, &manifest->vm[i], vm_id));
	}

	if (!found_primary_vm) {
		return MANIFEST_ERROR_NO_PRIMARY_VM;
	}

	return MANIFEST_SUCCESS;
}

const char *manifest_strerror(enum manifest_return_code ret_code)
{
	switch (ret_code) {
	case MANIFEST_SUCCESS:
		return "Success";
	case MANIFEST_ERROR_NO_HYPERVISOR_FDT_NODE:
		return "Could not find \"hypervisor\" node in manifest";
	case MANIFEST_ERROR_NOT_COMPATIBLE:
		return "Hypervisor manifest entry not compatible with Hafnium";
	case MANIFEST_ERROR_RESERVED_VM_ID:
		return "Manifest defines a VM with a reserved ID";
	case MANIFEST_ERROR_NO_PRIMARY_VM:
		return "Manifest does not contain a primary VM entry";
	case MANIFEST_ERROR_TOO_MANY_VMS:
		return "Manifest specifies more VMs than Hafnium has "
		       "statically allocated space for";
	case MANIFEST_ERROR_PROPERTY_NOT_FOUND:
		return "Property not found";
	case MANIFEST_ERROR_MALFORMED_STRING:
		return "Malformed string property";
	case MANIFEST_ERROR_STRING_TOO_LONG:
		return "String too long";
	case MANIFEST_ERROR_MALFORMED_STRING_LIST:
		return "Malformed string list property";
	case MANIFEST_ERROR_MALFORMED_INTEGER:
		return "Malformed integer property";
	case MANIFEST_ERROR_INTEGER_OVERFLOW:
		return "Integer overflow";
	case MANIFEST_ERROR_MALFORMED_INTEGER_LIST:
		return "Malformed integer list property";
	case MANIFEST_ERROR_MALFORMED_BOOLEAN:
		return "Malformed boolean property";
	}

	panic("Unexpected manifest return code.");
}
